design/design-doc-composition-functions.md

```diff
@@ -175,32 +179,557 @@ metadata:
   name: example
 spec:
   compositeTypeRef:
-    apiVersion: database.example.org/v1alpha1
+    apiVersion: database.example.org/v1
     kind: XPostgreSQLInstance
+  # This Composition specifies functions instead of (P&T) resources.
+  # This is an array - you can specify a pipeline of Functions.
   functions:
-  - name: my-cool-function
-    type: Container
-    container:
-      image: xkpg.io/my-cool-function:0.1.0
+    # Each entry in the functions array has a name. This uniquely identifies
+    # the Function 'call' within the Composition.
+  - name: compose-xr-using-go-templates
```

Comment:
I don't feel great about having `name` here and under `functionRef`, i.e. `functions[*].name` vs ``functions[*].functionRef.name`. It is not obvious that the first one is the name of the Function 'call' as commented above.

Wondering if we can improve this by using something else, like, `step`, `call`, `action`, etc., instead of `name`.

Reply:
Hmm, indeed we might even consider renaming `functions:` to something else, like `pipeline:`

```yaml
  pipeline:
  - step: compose-xr-using-go-templates
    functionRef:
      name: go-templates
    config:
      apiVersion: example.org/v1
      kind: GoTemplate
      source: Remote
      remote: git://github.com/example/my-xpostgresql-go-templates
```

Reply:
I like this API. Will update. Thanks!

Reply:
> there is a guarantee that each function in this list is called in the exact order they are listed?

Yes there is.

> I probably prefer keeping the name functions: actually

I mostly like the `step` over yet-another `name`. Right now I slightly prefer `pipeline` too but not strongly.